lab_3/main.py

```diff
@@ -12,33 +12,184 @@
+            if gram in self.gram_frequencies:
+                self.gram_frequencies[gram] += 1
+            else:
+                self.gram_frequencies[gram] = 1
+        return 'OK'
 
     def calculate_log_probabilities(self):
-        pass
+        if self.size == 2:
+            for gram, freq in self.gram_frequencies.items():
+                sum_of_freqs = 0 # gram from the same letter
+                for gram_2, freq_2 in self.gram_frequencies.items():
+                    if gram[0] == gram_2[0]:
+                        sum_of_freqs += freq_2
+                    else:
+                        continue
+                gram_prob = math.log(freq / sum_of_freqs)
+                self.gram_log_probabilities[gram] = gram_prob
+        if self.size == 3:
+            for gram, freq in self.gram_frequencies.items():
+                sum_of_freqs = 0 # gram from the same letter
+                for gram_2, freq_2 in self.gram_frequencies.items():
+                    if (gram[0] == gram_2[0]) and (gram[1] == gram_2[1]):
+                        sum_of_freqs += freq_2
+                    else:
+                        continue
+                gram_prob = math.log(freq / sum_of_freqs)
+                self.gram_log_probabilities[gram] = gram_prob
+
+        return 'OK'
 
     def predict_next_sentence(self, prefix: tuple) -> list:
-        pass
+        if not isinstance(prefix, tuple):
+            return []
+        if len(prefix) != self.size - 1:
+            return []
+        sentence = list()
+        if self.size == 2:
+            word = prefix[0]
+            sentence.append(word)
+            while True:
+                next_gram = list()
+                for gram, gram_prob in self.gram_log_probabilities.items():
+                    if gram[0] == word:
+                        next_gram.append((gram_prob, gram))
+
+                if len(next_gram) == 0:
+                    return sentence
+
+                next_gram.sort(reverse=True)
+                sentence.append(next_gram[0][1][1])
+                word = next_gram[0][1][1]
+
+        if self.size == 3:
+            words = [prefix[0], prefix[1]]
+            sentence = [prefix[0], prefix[1], ]
+            while True:
+                next_gram = list()
+                for gram, gram_prob in self.gram_log_probabilities.items():
+                    if (gram[0] == words[0]) and (gram[1] == words[1]):
+                        next_gram.append((gram_prob, gram))
+
+                if len(next_gram) == 0:
+                    return sentence
+
+                next_gram.sort(reverse=True)
+                sentence.append(next_gram[0][1][2])
+                words[0] = next_gram[0][1][1]
+                words[1] = next_gram[0][1][2]
 
 
 def encode(storage_instance, corpus) -> list:
-    pass
+    encoded = list()
+    for sentence in corpus:
+        lst = list()
+        for word in sentence:
+            lst.append(storage_instance[word])
```

Comment:
@irenemfzlv storage_instance is not a dict, its WordStorage object
called in the following context

```python
ws = WordStorage()
...
encoded = encode(ws, sentences)
...
```